presto-main/src/main/java/com/facebook/presto/memory/HighMemoryTaskKiller.java

```diff
@@ -0,0 +1,248 @@
+                continue;
+            }
+
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().addNotificationListener(
+                        objectName,
+                        gcNotificationListener,
+                        null,
+                        null);
+            }
+            catch (JMException e) {
+                throw new RuntimeException("Unable to add listener", e);
+            }
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (!taskKillerEnabled) {
+            return;
+        }
+
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().removeNotificationListener(objectName, gcNotificationListener);
+            }
+            catch (JMException ignored) {
+                log.error("Error removing notification: " + ignored);
+            }
+        }
+    }
+
+    private void onGCNotification(Notification notification)
+    {
+        if (GC_NOTIFICATION_TYPE.equals(notification.getType())) {
+            GarbageCollectionNotificationInfo info = new GarbageCollectionNotificationInfo((CompositeData) notification.getUserData());
+            if (info.isMajorGc()) {
+                if (shouldTriggerTaskKiller(info)) {
+                    //Kill task consuming most memory
+                    List<SqlTask> activeTasks = getActiveTasks();
+                    ListMultimap<QueryId, SqlTask> activeQueriesToTasksMap = activeTasks.stream()
+                            .collect(toImmutableListMultimap(task -> task.getQueryContext().getQueryId(), Function.identity()));
+
+                    Optional<QueryId> queryId = getMaxMemoryConsumingQuery(activeQueriesToTasksMap);
+
+                    if (queryId.isPresent()) {
+                        List<SqlTask> activeTasksToKill = activeQueriesToTasksMap.get(queryId.get());
+                        for (SqlTask sqlTask : activeTasksToKill) {
+                            TaskStats taskStats = sqlTask.getTaskInfo().getStats();
+                            sqlTask.failed(new PrestoException(EXCEEDED_HEAP_MEMORY_LIMIT, format("Worker heap memory limit exceeded: User Memory: %d, System Memory: %d, Revocable Memory: %d", taskStats.getUserMemoryReservationInBytes(), taskStats.getSystemMemoryReservationInBytes(), taskStats.getRevocableMemoryReservationInBytes())));
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private boolean shouldTriggerTaskKiller(GarbageCollectionNotificationInfo info)
+    {
+        boolean triggerTaskKiller = false;
+        DataSize beforeGcDataSize = info.getBeforeGcTotal();
+        DataSize afterGcDataSize = info.getAfterGcTotal();
+
+        if (taskKillerStrategy == FREE_MEMORY_ON_FREQUENT_FULL_GC) {
```

Comment:
Should this be called FREE_MEMORY_... or just TRIGGER_ON_ or even KILL_ON_... ? It doesn't free any memory explicitly other than failing queries.